File: 0312_burst_balloons/Solution.cs
// LeetCode 0312 - Burst Balloons

// https://leetcode.com/problems/burst-balloons/



public class Solution {

    public int MaxCoins(int[] nums) {

        int size = nums.Length + 2;

        int[] balloons = new int[size];

        balloons[0] = 1;

        balloons[size - 1] = 1;

        for (int index = 0; index < nums.Length; index++) {

            balloons[index + 1] = nums[index];

        }



        int[,] dp = new int[size, size];

        for (int length = 3; length <= size; length++) {

            for (int left = 0; left <= size - length; left++) {

                int right = left + length - 1;

                for (int mid = left + 1; mid < right; mid++) {

                    int coins = dp[left, mid] + dp[mid, right]

                        + balloons[left] * balloons[mid] * balloons[right];

                    dp[left, right] = System.Math.Max(dp[left, right], coins);

                }

            }

        }

        return dp[0, size - 1];

    }

}

File: 0312_burst_balloons/Solution.scala
// LeetCode 0312 - Burst Balloons

// https://leetcode.com/problems/burst-balloons/



object Solution {

  def maxCoins(nums: Array[Int]): Int = {

    val balloons = Array(1) ++ nums ++ Array(1)

    val size = balloons.length

    val dp = Array.fill(size, size)(0)

    for (length <- 3 to size) {

      for (left <- 0 to size - length) {

        val right = left + length - 1

        for (mid <- left + 1 until right) {

          val coins = dp(left)(mid) + dp(mid)(right) +

            balloons(left) * balloons(mid) * balloons(right)

          dp(left)(right) = math.max(dp(left)(right), coins)

        }

      }

    }

    dp(0)(size - 1)

  }

}

